pkg/config/spec.go

```diff
@@ -265,30 +265,9 @@ func NewUpgradeSpec(cfg *Config) (*v1.UpgradeSpec, error) {
 		State:      installState,
 	}
 
-	source := viper.GetString("upgradeSource")
-	recoveryUpgrade := viper.GetBool("upgradeRecovery")
-	if source != "" {
-		imgSource, err := v1.NewSrcFromURI(source)
-		if err == nil {
-			if recoveryUpgrade {
-				spec.RecoveryUpgrade = recoveryUpgrade
-				spec.Recovery.Source = imgSource
-			} else {
-				spec.Active.Source = imgSource
-			}
-			size, err := GetSourceSize(cfg, imgSource)
-			if err != nil {
-				cfg.Logger.Warnf("Failed to infer size for images: %s", err.Error())
-			} else {
-				cfg.Logger.Infof("Setting image size to %dMb", size)
-				// On upgrade only the active or recovery will be upgraded, so we dont need to override passive
-				if recoveryUpgrade {
-					spec.Recovery.Size = uint(size)
-				} else {
-					spec.Active.Size = uint(size)
-				}
-			}
-		}
+	err = setSourceSize(cfg, spec)
```

Comment:
This is too early to run this. We haven't unmarshalled with yip yet thus spec.Recovery.Source (or spec.Active.Source) is not set yet.

Reply:
refactored code to address this. Now we unmarshal the yaml before we calculate sizes.